src/transformers/generation_utils.py

```diff
@@ -1645,6 +1759,330 @@ def typeerror():
+
+        # keep track of which sequences are already finished
+        unfinished_sequences = input_ids.new(input_ids.shape[0]).fill_(1)
+
+        this_peer_finished = False  # used by synced_gpus only
+
+        step_counter = 0
+        while True:
+            if synced_gpus:
+                # Under synced_gpus the `forward` call must continue until all gpus complete their sequence.
+                # The following logic allows an early break if all peers finished generating their sequence
+                this_peer_finished_flag = torch.tensor(0.0 if this_peer_finished else 1.0).to(input_ids.device)
+                # send 0.0 if we finished, 1.0 otherwise
+                dist.all_reduce(this_peer_finished_flag, op=dist.ReduceOp.SUM)
+                # did all peers finish? the reduced sum will be 0.0 then
+                if this_peer_finished_flag.item() == 0.0:
+                    break
+
+            # if the first step in the loop, encode all the prefix and obtain three parameters: (1) past_key_values; (2) last_hidden_states; (3) logit_for_next_step
+            if step_counter == 0:
+                # encode the given prefix and prepare model inputs; encoder-decoder model process the prefix and save the `encoder_outputs`
+                model_kwargs["use_cache"] = True
+                model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
+                output = self(**model_inputs, output_hidden_states=True, output_attentions=True)
+                # past_key_values is activated for fast decoding
+                past_key_values = output.past_key_values
+                # last decoder hidden states will be used to compute the degeneration penalty (cosine similarity with previous tokens)
+                if self.config.is_encoder_decoder:
+                    last_hidden_states = output.decoder_hidden_states[-1]
+                else:
+                    last_hidden_states = output.hidden_states[-1]
+                # next logit for contrastive search to select top-k candidate tokens
+                logit_for_next_step = output.logits[:, -1, :]
+
+            # contrastive_search main logic start:
+            # contrastive search decoding consists of two steps: (1) candidate tokens recall; (2) candidate re-rank by degeneration penalty
+            bsz, seqlen, embed_dim = last_hidden_states.size()
+
+            # logits processor: empty logits processor
+            logit_for_next_step = logits_processor(input_ids, logit_for_next_step)
+            next_probs = nn.functional.softmax(logit_for_next_step, dim=-1)
+            # logit warper: applying after the softmax, which is consistent with the logic in the paper
+            logit_for_next_step = logits_warper(input_ids, next_probs)
+
+            _, top_k_ids = torch.topk(logit_for_next_step, dim=-1, k=top_k)
+            top_k_probs = torch.gather(next_probs, dim=1, index=top_k_ids)
```

Comment:
Uhmmm I see. Because the top k IDs are needed, and our `logits_warper` only returns the updated scores, a `torch.topk()` call outside the logits warper is always needed 🤔 The `logits_warper` already performs a `top_k` call, but it only filters the scores. However, `logits_warper` may also include other operations (e.g. nucleus sampling-like filtering, with `top_p`), so I think it's positive to be included here.

A potential change might be needed, though: `logits_warper` can change the logits, so the softmax should happen after that line.